gateway-server/src/main/java/org/apache/knox/gateway/services/AbstractGatewayServices.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.knox.gateway.config.GatewayConfig;
+import org.apache.knox.gateway.i18n.messages.MessagesFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * AbstractGatewayServices is an abstract implementation of {@link GatewayServices} that manages the
+ * contained/registered services. See {@link ServiceType} for the different types of services that may
+ * be registered.
+ * <p>
+ * This implementation ensures the proper ordering of registered services when starting and stopping
+ * them.
+ * <p>
+ * {@link GatewayServices} implementations should extend this class and register relevant services
+ * during the initialization process ({@link GatewayServices#init(GatewayConfig, Map)} using
+ * {@link #addService(ServiceType, Service)}.
+ * <p>
+ */
+public abstract class AbstractGatewayServices implements GatewayServices {
+
+  private static final GatewayMessages LOG = MessagesFactory.get(GatewayMessages.class);
+
+  private final Map<ServiceType, Service> services = new EnumMap<>(ServiceType.class);
+  private final List<Service> orderedServices = new ArrayList<>();
+  private final String role;
+  private final String name;
+
+  AbstractGatewayServices(String role, String name) {
+    this.role = role;
+    this.name = name;
+  }
+
+  @Override
+  public void start() throws ServiceLifecycleException {
+    for (Service service : orderedServices) {
+      LOG.startingService(service.getClass().getName());
+      service.start();
+    }
+  }
+
+  @Override
+  public void stop() throws ServiceLifecycleException {
+    // Reverse the preferred startup order
+    ArrayList<Service> copy = new ArrayList<>(orderedServices);
+    Collections.reverse(copy);
+
+    for (Service service : copy) {
+      LOG.stoppingService(service.getClass().getName());
+      service.stop();
+    }
+  }
+
+  @Override
+  public String getRole() {
+    return role;
+  }
+
+  @Override
+  public String getName() {
+    return name;
+  }
+
+  @Override
+  public Set<ServiceType> getServiceTypes() {
+    return services.keySet();
+  }
+
+  @Override
+  public <T> T getService(ServiceType serviceType) {
+    return (T) services.get(serviceType);
+  }
+
+  /**
+   * Add/register a service with this container.
+   * <p>
+   * An implementation for a service type may be added/registered only once.  Adding a service
+   * implementation for a service type that was previously added overwrites the previously added
+   * implementation.
+   *
+   * @param serviceType the {@link ServiceType} type to add
+   * @param service     the {@link Service} implementation
+   */
+  protected void addService(ServiceType serviceType, Service service) {
+
+    // Ensure the service is not null
+    if (service == null) {
+      throw new NullPointerException("A null service may not be added.");
+    }
+
+    orderedServices.add(service);
```

Comment:
I think we should make sure we don't add the same service twice. This would case the same service to be started twice which could cause issues potentially.

Reply:
It may be possible for the same class to implement multiple services.  So it is conceivable that a class instance could be added for multiple types.  That said, may be lifecycle methods (init, start, stop) should be smarter and require a certain state to exist before being allowed to be executed.  

For example:
* you should not be able to `start` a service that was previously started (warn, and ignore?)
* you should not be able to `init` a started service.  (throw exception?)

Reply:
Hmm thats a good point. That would be a bigger change though.